src/chainstate/stacks/boot/pox.clar

```diff
@@ -0,0 +1,754 @@
+;; for which its clients' uSTX will be locked.
+;; tx-sender is the delegate.
+(define-public (register-delegate (pox-addr (tuple (version uint) (hashbytes (buff 20))))
+                                  (tenure-burn-block-begin uint)
+                                  (tenure-reward-num-cycles uint))
+    (let (
+        ;; tenure begins in the first full reward cycle after when the tenure burn block passes
+        (first-allowed-reward-cycle 
+            (+ u1 (burn-height-to-reward-cycle tenure-burn-block-begin)))
+    )
+    (begin
+        ;; must not be a registered delegate already
+        (asserts! (is-none (get-delegate-control-info tx-sender))
+            (err ERR-STACKING-DELEGATE-ALREADY-REGISTERED))
+
+        ;; lock period must be in acceptable range.
+        (asserts! (check-pox-lock-period tenure-reward-num-cycles)
+            (err ERR-STACKING-INVALID-DELEGATE-TENURE))
+
+        ;; PoX address version must be valid
+        (asserts! (check-pox-addr-version (get version pox-addr))
+            (err ERR-STACKING-INVALID-POX-ADDRESS))
+
+        ;; tenure must start strictly before the first-allowed reward cycle
+        (asserts! (< tenure-burn-block-begin (reward-cycle-to-burn-height first-allowed-reward-cycle))
+            (err ERR-STACKING-INVALID-DELEGATE-TENURE))
+
+        ;; tenure must start at or after the current burn block height -- no retroactive registration
+        (asserts! (<= burn-block-height tenure-burn-block-begin)
+            (err ERR-STACKING-INVALID-DELEGATE-TENURE))
+
+        ;; register!
+        (map-set delegate-control
+            { delegate: tx-sender }
+            {
+                total-ustx: u0,
+                pox-addr: pox-addr,
+                burn-block-height-start: tenure-burn-block-begin,
+                first-reward-cycle: first-allowed-reward-cycle,
+                tenure-cycles: tenure-reward-num-cycles
+            }
+        )
+
+        ;; claim PoX address up-front
+        (map-set pox-addr-reward-cycles
+            { pox-addr: pox-addr }
+            { first-reward-cycle: first-allowed-reward-cycle, num-cycles: tenure-reward-num-cycles }
+        )
+        (ok true)
+    ))
+)
+
+;; Delegated uSTX lock-up.
+;; The Stacker (the caller) locks up their uSTX, and specifies a delegate
+;; configuration to use.  The delegate will then carry out the PoX address
+;; registration.
+;; Some checks are performed:
+;; * The Stacker must not have rejected PoX in the upcoming reward cycle
+;; * The Stacker must not currently be Stacking.
+;; * The Stacker must have the given amount-ustx funds available.
+;; * The Stacker cannot be a delegate
+;; * The stacker cannot have a delegate
+;; * The delegate's tenure must not have started
+;; * The delegate must have not have begun.
+;; The tokens will unlock and be returned to the Stacker (tx-sender) automatically once the
+;; lockup period ends.
+(define-public (delegate-stx (delegate principal)
```

Comment:
This delegation system behaves pretty differently from the delegation described in SIP-007, the biggest difference is that the delegator should be able to control the STX lockup for the provided address (such that they can re-lock after a lockup expires).

Reply:
Right now the delegate can only lock _once_.  And even then, it's not truly locking them; it's just committing the PoX address to the reward set to represent the aggregate STX balance.  The individuals who call `delegate-stx` lock their STX right then and there, and the delegate commits them to the reward set later once everyone has done so.

(This introduces a failure mode where a delegate can fail to commit in the allotted time frame, which would leave the users' STX locked up but not earning any BTC.  But, there's no real remediation option available to us within the Stacks node, short of limiting the number of STX holders that can be represented by a delegate.  Trying to do otherwise will re-introduce the griefing DDoS vector described above -- a lot of tiny Stackers will need to have their lock-ups reversed by the node.  In practice, I expect that this failure mode will be proactively prevented by smart contracts, which themselves can act as delegates, and can do things like escrow your STX instead of locking them up early, and can allow any participant to force the delegate to commit once an escrowed STX threshold has been met).

I thought that was the desired behavior?  i.e. Stacking requires some degree of continuous participation on behalf of the STX owners?

Reply:
I had a different impression of the design of delegation. One of the essential use cases of delegation is support for separating the action of Stacking from the keys owning the Stacks in question -- e.g., if Alice wants to participate in Stacking, but keeps her STX wallet in cold storage, delegation provides a mechanism to safely participate in Stacking without needing to continuously re-access the STX wallet.

Reply:
Thinking a little harder about this failure mode I described above, I think we need to remove this constraint from SIP-007:

```
Both types of delegation transactions must be signed by the represented address.
These are transactions on the Stacks blockchain, and will be implemented via a native
smart contract, loaded into the blockchain during the Stacks 2.0 genesis block. These
transactions, therefore, are contract-call invocations. The invoked methods are guarded
by:

    (asserts! (is-eq contract-caller tx-sender) (err u0))

This insures that the methods can only be invoked by direct transaction execution.
```

If we had this constraint, then I don't think there's a way to avoid a delegate screwing over its users by refusing to commit its clients' STX to a reward address.  This is because users locking their STX for a delegate is a separate step from the delegate committing them to a reward address.  I think these two actions _need_ to be done separately -- one by the user and one by the delegate -- because we can't have users naively giving their STX to a delegate in order to lock them (otherwise, the delegate could just steal them).  The user either needs to programmatically constrain what a delegate can do with their imminently-locked STX (which in turn requires that Stacking be carried out by a smart contract, in violation of the above SIP-007 text), or the user needs to lock their STX before the delegate takes any action (which opens up the aforementioned failure mode).

Reply:
> I had a different impression of the design of delegation. One of the essential use cases of delegation is support for separating the action of Stacking from the keys owning the Stacks in question -- e.g., if Alice wants to participate in Stacking, but keeps her STX wallet in cold storage, delegation provides a mechanism to safely participate in Stacking without needing to continuously re-access the STX wallet.

This can still be done with the current design.  If we remove the above text in SIP-007, then the user could put their STX into a smart contract that acts as a "stacking vault."  The user would have three keys -- a cold-storage key that empties the vault, a hot-storage key that re-stacks their tokens with a delegate, and a delegate-rotation key that de-registers the delegate's hot-storage key and sets a new one.  The delegate would be given the re-stacking key, and if the user wanted to switch delegates, the user uses the delegate-rotation key to do so.  The tokens could only be removed from the "stacking vault" contract with the cold-storage key.

Reply:
I think that it might make sense to separate delegation from this PR -- I think there's a number of decisions related to the workings of delegation that deserve to have a larger discussion, and that process shouldn't also block the implementation of the locking mechanism (and integration of that locking into sortition).

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/1850/commits/d97b5842ff9f902f9d1c84f43c7101455e848ca4